crates/chia-consensus/src/merkle_tree.rs

```diff
@@ -0,0 +1,1362 @@
+        Err(SetError)
+    } else {
+        Ok(merkle_tree)
+    }
+}
+
+fn _deserialize(merkle_tree: &mut MerkleSet, proof: &[u8], pos: usize) -> Result<usize, SetError> {
+    if let Some(&t) = proof.get(pos) {
+        match t {
+            EMPTY => {
+                merkle_tree.nodes_vec.push(ArrayTypes::Empty);
+                merkle_tree.hash_cache.push(BLANK);
+                Ok(pos + 1)
+            }
+            TERMINAL => {
+                let hash: [u8; 32] = proof[pos + 1..pos + 33].try_into().map_err(|_| SetError)?;
+                merkle_tree.leaf_vec.push(hash);
+                merkle_tree.nodes_vec.push(ArrayTypes::Leaf {
+                    data: merkle_tree.leaf_vec.len() - 1,
+                });
+                merkle_tree.hash_cache.push(hash);
+                Ok(pos + 33)
+            }
+            TRUNCATED => {
+                let hash: [u8; 32] = proof[pos + 1..pos + 33].try_into().map_err(|_| SetError)?;
+                merkle_tree.nodes_vec.push(ArrayTypes::Truncated);
+                merkle_tree.hash_cache.push(hash);
+                Ok(pos + 33)
+            }
+            MIDDLE => {
+                let new_pos = _deserialize(merkle_tree, proof, pos + 1)?;
+                let left_pointer = merkle_tree.nodes_vec.len() - 1;
+
+                let final_pos = _deserialize(merkle_tree, proof, new_pos)?;
+                let right_pointer = merkle_tree.nodes_vec.len() - 1;
+                merkle_tree.nodes_vec.push(ArrayTypes::Middle {
+                    children: (left_pointer, right_pointer),
+                });
+                let left_type = array_type_to_node_type(merkle_tree.nodes_vec[left_pointer]);
+                let right_type = array_type_to_node_type(merkle_tree.nodes_vec[right_pointer]);
+                let node_hash = hash(
+                    left_type,
+                    right_type,
+                    &merkle_tree.hash_cache[left_pointer],
+                    &merkle_tree.hash_cache[right_pointer],
+                );
+                merkle_tree.hash_cache.push(node_hash);
+                Ok(final_pos)
+            }
+            _ => Err(SetError),
+        }
+    } else {
+        Err(SetError)
+    }
+}
+
+impl Default for MerkleSet {
+    fn default() -> MerkleSet {
+        MerkleSet {
+            nodes_vec: Vec::new(),
+            leaf_vec: Vec::new(),
+            hash_cache: Vec::new(),
+        }
+    }
+}
+
+impl MerkleSet {
+    pub fn get_merkle_root(&self) -> [u8; 32] {
+        self.hash_cache[self.hash_cache.len() - 1]
+    }
+
+    // returns a tuple of a bool representing if the value has been found, and if so bytes that represent the proof of inclusion
+    pub fn generate_proof(&self, included_leaf: [u8; 32]) -> Result<(bool, Vec<u8>), SetError> {
+        let mut proof = Vec::new();
+        let r = self.is_included(self.nodes_vec.len() - 1, included_leaf, &mut proof, 0)?;
+        Ok((r, proof))
+    }
+
+    pub fn is_included(
```

Comment:
Again, structure / name brought over from Python

Reply:
this isn't exported to python though, so I think it making sense is more important. But maybe there's a really good reason to call it this (it would warrant a comment though, I think). The convention in rust is that functions of the form `is_*()` query the state of a class.